kv/memberlist/tcp_transport.go

```diff
@@ -376,24 +380,45 @@ func (t *TCPTransport) FinalAdvertiseAddr(ip string, port int) (net.IP, int, err
 				return nil, 0, fmt.Errorf("no private IP address found, and explicit IP not provided")
 			}
 
-			advertiseAddr = net.ParseIP(ip)
-			if advertiseAddr == nil {
-				return nil, 0, fmt.Errorf("failed to parse advertise address: %q", ip)
+			addr, err := netip.ParseAddr(ip)
+			if err != nil {
+				return nil, 0, fmt.Errorf("failed to parse advertise address %q", ip)
 			}
-		} else {
+
+			if addr.IsValid() {
+				advertiseAddr = addr
+			}
+		case colonColonZero:
+			inet6Ip, err := netutil.GetFirstAddressOf([]string{}, t.logger, true)
+			if err != nil {
+				return nil, 0, fmt.Errorf("failed to get private inet6 address: %w", err)
+			}
+
+			addr, err := netip.ParseAddr(inet6Ip)
+			if err != nil {
+				return nil, 0, fmt.Errorf("failed to parse inet6 advertise address %q", ip)
+			}
+			advertiseAddr = addr
+		default:
 			// Use the IP that we're bound to, based on the first
 			// TCP listener, which we already ensure is there.
-			advertiseAddr = t.tcpListeners[0].Addr().(*net.TCPAddr).IP
+			addr, err := netip.ParseAddr(t.tcpListeners[0].Addr().(*net.TCPAddr).IP.String())
```

Comment:
I am curious about this.  Why isn't it always acceptable to parse the TCP listener?  If we're bound to an interface for one part of the code, shouldn't we be using that for everything else?

Reply:
The answer is that the socket may be 0.0.0.0, which we don't want to advertise to peers.